include/rosidl_dynamic_typesupport/api/dynamic_data.h

```diff
@@ -22,12 +22,39 @@
 extern "C" {
 #endif
 
+#include <rcutils/allocator.h>
 #include <rcutils/types/rcutils_ret.h>
 #include <rcutils/types/uint8_array.h>
-#include <rosidl_dynamic_typesupport/api/serialization_support_interface.h>
-#include <rosidl_dynamic_typesupport/visibility_control.h>
-#include <rosidl_dynamic_typesupport/uchar.h>
 
+#include "rosidl_dynamic_typesupport/api/serialization_support.h"
+#include "rosidl_dynamic_typesupport/api/serialization_support_interface.h"
+#include "rosidl_dynamic_typesupport/visibility_control.h"
+#include "rosidl_dynamic_typesupport/types.h"
+#include "rosidl_dynamic_typesupport/uchar.h"
+
+// Dynamic Data Impl
+struct rosidl_dynamic_typesupport_dynamic_data_impl_s
+{
+  rcutils_allocator_t allocator;
+  void * handle;
+};
+
+ROSIDL_DYNAMIC_TYPESUPPORT_PUBLIC
+rosidl_dynamic_typesupport_dynamic_data_impl_t
+rosidl_dynamic_typesupport_get_zero_initialized_dynamic_data_impl(void);
+
+// Dynamic Data
+struct rosidl_dynamic_typesupport_dynamic_data_s
+{
+  rcutils_allocator_t allocator;
+  rosidl_dynamic_typesupport_dynamic_data_impl_t impl;
```

Comment:
Not using hidden-type `pimpl`?

Reply:
I was thinking of doing that, but I thought it would be better to enforce that the impl is a complete type (and then defer it to the handle), so that we can enforce that the impl type has an allocator

Reply:
But isn't the the type defined just above?  That is, this implementation is completely in control of that structure, and thus can guarantee that it always has an allocator, no?

Reply:
ah, right.. I think I'm avoiding the use of a pointer so we don't have to keep doing null checks on the impl member as well..

The impl is unusable without the associated serialization support in either case, so in my view composition makes sense

Reply:
Yeah, actually, I largely agree with you.  The only problem here is that if we want to add, remove, or change structure members in the implementation, we won't be able to backport it due to ABI concerns.  But if we are OK with that limitation for now, we can leave it as-is.

Reply:
The implementation contains a void * handle, so that's a good escape hatch I think?

Reply:
Yeah, it's good enough for now.  It isn't our usual style, but it will do in a pinch.